MycroftAI/mycroft-core
# Ability to play sound instead of utterance

==== Fixed Issues ====
issue-2755 -  Ability to play sound instead of an utterance

PULL REQUEST EDITED TO MATCH SOME COMMENTS : this is now a broader PR, with the ability to use sound for all "final" dialog, and not only error one. Several sounds could be used (all those listed in the "sounds" parameter)
This PR is also extended to allow skill developper to use their own sound

====  Tech Notes ====
- Add a new method report_outcome for skills to declare if the utterance passed in parameter is final. For the parameter ("outcome_type"), the possible values are constrained by those available in the "sounds" dictionnary parameter in mycroft.conf.
    With the second parameter "custom_sound" (path to a wav or mp3),  skills can use their own sound, and not only the basic "sounds" from Mycroft. Note : I thought about merging the two parameters, but I think they serve different purposes (one is more a meta information about the utterance, the other is a custom sound for broader use). But I could change this, of course.
    Add corresponding new method parameters in the speak method

- How Mycroft choose between a sound and the utterance with the report_outcome method  ? : 
.. If the utterance is empty (None), the sound will be used. It's the way for the skill developper to force sound.
.. If uterrance is set and outcome_type or custom_sound is set, then Mycroft will check the new parameters (skills_pref_sound to check if the current skills is in the list, or always_pref_sound for a global parameter). If one of this parameter matches, then the sound will be used.
.. In all other cases, the uterrance will be used.

- Add a new error sound. (quite horrible, sorry. It's the wakeword sound, reversed)

- Change some TTS : they originally overrode a method they should not, in my opinion. If my guess is good they should override the inner "_execute" method and not the main public one. I could need some educated guess from the reviewer for this point.

- Implementation choice, not sure of my choice : The sound is played in the TTS module. It could be played upper in the stack but I did this in order to keep the functionality close to its replacement. And also using the queue seems to me the proper way to replace an uterrance.

TODO ? : 
- The "dont_understand" sound I made is horrible. If someone has a better one... Please be my guest !
- in the "sounds" parameter list available by default, one sound doesn't seem to be in the filesystem ('end_listening')
- To add other sounds to complete the range of choices.


====  Documentation Notes ====
Two new parameters can force Mycroft to do feedback with a brief sound instead of a complete utterance :
skills.skills_pref_sound: a list of skills who should use sound. Use the class name.
always_pref_sound: boolean to force Mycroft to use sound for feedback, wherever it is possible.
Skill developers must use the "report_outcome" method, or the speak method with the "outcome_type" or "custom_sound" to use this functionality

==== Localization Notes ====
NONE

==== Environment Notes ====
NONE

==== Protocol Notes ====
NONE

## Description
fixes #issue-2755

## How to test
Modify a skill (for example, fallback-unknown), to use "report_outcome" instead of speak_dialog. Or use "outcome_type"/"custom_sound" parameter with the speak method
Example :
`self.speak('unknown', outcome_type='dont_understand')`
`self.report_outcome('dialog', outcome_type='dont_understand')`
Activate the replacement sound for this skill, either by the global configuration parameter (always_pref_sound) or by white listing the "UnknownSkill" in the list parameter skills.skills_pref_sound (class name)
Make mycroft says this dialog in fallback-unknown by asking him something nonsensical.
Instead of the unknown dialog, Mycroft should play 'dont_understand' the sound.

Do the same but with the parameter 'custom_sound' to play any sound on the file system.

## Contributor license agreement signed?
CLA [X]
